File: internal/interfaces/services/mdns/mdns.go
// SPDX-License-Identifier: BSD-2-Clause
//
// Copyright (c) 2025 The FreeBSD Foundation.
//
// This software was developed by Hayzam Sherif <hayzam@alchemilla.io>
// of Alchemilla Ventures Pvt. Ltd. <hello@alchemilla.io>,
// under sponsorship from the FreeBSD Foundation.

package mdnsInterfaces

import (
	"github.com/alchemillahq/sylve/internal/db/models/mdns"
)

type MdnsRecordWithManaged struct {
	mdnsModels.MdnsRecord
	Managed bool   `json:"managed"`
	Source  string `json:"source"`
}

type MdnsServiceInterface interface {
	Rebuild() error
	GetSettings() (mdnsModels.MdnsSettings, error)
	SetSettings(interfaces, hostname string) error
	GetRecords() ([]MdnsRecordWithManaged, error)
	CreateRecord(name, recordType string, port int, txt map[string]string, interfaces string) (mdnsModels.MdnsRecord, error)
	UpdateRecord(id uint, name, recordType string, port int, txt map[string]string, interfaces string) error
	DeleteRecord(id uint) error
}

File: web/src/lib/types/network/mdns.ts
import { z } from 'zod/v4';

export const MdnsSettingsSchema = z.object({
    id: z.number(),
    interfaces: z.string().default(''),
    hostname: z.string().default('')
});

export type MdnsSettings = z.infer<typeof MdnsSettingsSchema>;

export const MdnsRecordSchema = z.object({
    id: z.number(),
    name: z.string(),
    type: z.string(),
    port: z.number(),
    txt: z.record(z.string(), z.string()).nullish().transform((value) => value ?? {}),
    interfaces: z.string().default(''),
    createdAt: z.string(),
    updatedAt: z.string()
});

export type MdnsRecord = z.infer<typeof MdnsRecordSchema>;

export const MdnsRecordWithManagedSchema = MdnsRecordSchema.extend({
    managed: z.boolean(),
    source: z.string()
});

export type MdnsRecordWithManaged = z.infer<typeof MdnsRecordWithManagedSchema>;
